Kernel/RTC.cpp

```diff
@@ -70,12 +90,6 @@ void read_registers(unsigned& year, unsigned& month, unsigned& day, unsigned& ho
 
 time_t now()
 {
-    // FIXME: We should probably do something more robust here.
-    //        Perhaps read all the values twice and verify that they were identical.
-    //        We don't want to be caught in the middle of an RTC register update.
```

Comment:
This patch does not address the FIXME at all. The comment talks about reading the registers in the middle of them being updated, not about having a timeout.

Reply:
I understand that and I shouldn't remove this FIXME now that I see what it aims to, but the comment says:
```
We don't want to be caught in the middle of an RTC register update.
```
does it make sense to check twice that we are not in the middle of the RTC update? Also, the current code checks twice for being in the middle of update, before reading (and not before and after if it makes sense somehow). 
What the OSDev wiki suggests to do is:
```
To help guard against this problem the RTC has an "Update in progress" flag (bit 7 of Status Register A). To read the time and date properly you have to wait until the "Update in progress" flag goes from "set" to "clear". This is not the same as checking that the "Update in progress" flag is clear.

For example, if code does "while(update_in_progress_flag != clear)" and then starts reading all the time and date registers, then the update could begin immediately after the "Update in progress" flag was checked and the code could still read dodgy/inconsistent values. To avoid this, code should wait until the flag becomes set and then wait until the flag becomes clear. That way there's almost 1 second of time to read all of the registers correctly. 
```
The problem is that technique is slow (we waste about a second or two).
There are solutions to this as specified here:
https://wiki.osdev.org/CMOS#RTC_Update_In_Progress
Maybe there's no good way to determine if we have good values, and we should be careful when taking the RTC values into account with time-keeping.

Reply:
What if we simply read the values twice and confirm that the values are identical?

Reply:
How do you do that actually, if for example, you read the the time now, wait a bit and try again, but the minute counter went up, or the hour counter went up, or maybe you read it just before the end of the "24 hours" counter was reset because it's a new day? You subtract the values? what is considered a reasonable time diff? Do you keep reading the registers twice until you get identical values? what is an appropriate timeout for this procedure?

Also, I don't know how much the CMOS RTC is slow, but I bet that on real hardware it's slower than what the emulator exposes (due to reading from IO ports which tends to be slow sometimes), so, I might be wrong about this, but I don't see a way to get the same identical values if you read the registers twice. Anyway, nobody can guarantee this to work, even though the OSDev wiki says we could that.

Maybe I'm too pedantic about this, but if we want a solution for this, we need to consider these problems. Or maybe, again, we can be OK with dodgy values (so we try to read twice to get identical values but do have a reasonable timeout for not doing this forever), and let ntpupdate utility to update the time for us. Or maybe we should look at what other systems do and to draw a conclusion from this little research.

Reply:
>I don't see a way to get the same identical values if you read the registers twice that easily

We are reading the following values from CMOS:
* year
* month
* day
* hour
* minute
* second

It should be extremely possible to get identical values twice, unless you are suggesting that reading these values could take more than 1 second? :)